include/mun/invoke_result.h

```diff
@@ -0,0 +1,114 @@
+         * \param args arguments for the retry_fn
+         */
+        explicit InvokeResult(retry_fn_type&& retry_fn, update_fn_type&& update_fn, Args&&... args)
+            : m_variant(std::tuple(
+                std::move(retry_fn),
+                std::move(update_fn),
+                std::tuple(std::forward<Args>(args)...)
+            )) {}
+
+        /** Retrieves whether the function invocation succeeded. */
+        bool is_ok() noexcept {
+            return std::holds_alternative<success_type>(m_variant);
+        }
+
+        /** Retrieves whether the function invocation failed. */
+        bool is_err() noexcept {
+            return std::holds_alternative<err_type>(m_variant);
+        }
+
+        /** Retrieves the output of a successful function invocation.
+         *
+         * BEWARE: Calling this on a failed invocation result will result in undefined behavior.
+         *
+         * \return the output of a successful function invocation
+         */
+        success_type&& unwrap() noexcept {
+            return std::move(std::get<0>(m_variant));
+        }
+
+        /** Retrieves the error state of a failed function invocation.
+         *
+         * BEWARE: Calling this on a successful invocation result will result in undefined behavior.
+         *
+         * \return the error state
+         */
+        err_type&& unwrap_err() noexcept {
+            return std::move(std::get<1>(m_variant));
+        }
+
+        /** Retries a failed function invocation and returns the result, or immediately returns on prior success.
+         * 
+         * This will wait on updates from the `update_fn` callback before retrying.
+         * 
+         * \return the result of a retried function invocation
+         */
+        result_type&& retry() noexcept {
+            if (is_err()) {
+                return retry_impl(std::make_index_sequence<sizeof...(Args)>());
+            } else {
+                return std::move(*this);
+            }
+        }
+
+        /** Keeps retrying the function invocation until success, at which point it returns the function's output.
+         *
+         * BEWARE: This function will not terminate until the function has been successfully invoked.
+         *
+         * \return the output of a succesful function invocation
+         */
+        success_type&& wait() noexcept {
+            while (is_err()) {
+                *this = std::move(retry_impl(std::make_index_sequence<sizeof...(Args)>()));
+            }
+            return unwrap();
+        }
+
+    private:
+        template <std::size_t... Is>
+        InvokeResult<Output, Args...> retry_impl(std::index_sequence<Is...>) {
+            auto err = std::move(unwrap_err());
+            auto& retry_fn = std::get<0>(err);
+            auto& update_fn = std::get<1>(err);
+            auto& args = std::get<2>(err);
+            while (!update_fn()) {
+                // Spin until an update has occured that might fix the error
```

Comment:
Maybe at least put a std::this_thread::sleep in there? Or a yield?

Reply:
I've added `std::this_thread::sleep` for now and added a TODO for adding wait functionality inside the Rust Runtime API. This is non-trivial, so will have to wait for now.